libraries/tracker-core/src/emitter/index.ts

```diff
@@ -0,0 +1,410 @@
+  port?: number;
+  /* If the request should undergo server anonymization. */
+  serverAnonymization?: boolean;
+}
+
+/**
+ * Emitter is responsible for sending events to the collector.
+ * It manages the event queue and sends events in batches depending on configuration.
+ */
+export interface Emitter {
+  /**
+   * Forces the emitter to send all events in the event store to the collector.
+   * @returns A Promise that resolves when all events have been sent to the collector.
+   */
+  flush: () => Promise<void>;
+  /**
+   * Adds a payload to the event store or sends it to the collector.
+   * @param payload - A payload to be sent to the collector
+   * @returns Promise that resolves when the payload has been added to the event store or sent to the collector
+   */
+  input: (payload: Payload) => Promise<void>;
+  /**
+   * Updates the collector URL to which events will be sent.
+   * @param url - New collector URL
+   */
+  setCollectorUrl: (url: string) => void;
+  /**
+   * Sets the server anonymization flag.
+   */
+  setAnonymousTracking: (anonymous: boolean) => void;
+  /**
+   * Updates the buffer size of the emitter.
+   */
+  setBufferSize: (bufferSize: number) => void;
+}
+
+interface RequestResult {
+  success: boolean;
+  retry: boolean;
+  status?: number;
+}
+
+export function newEmitter({
+  endpoint,
+  eventMethod = 'post',
+  protocol,
+  port,
+  maxPostBytes = 40000,
+  maxGetBytes,
+  bufferSize = 1,
+  customHeaders,
+  serverAnonymization,
+  connectionTimeout,
+  keepalive,
+  idService,
+  dontRetryStatusCodes = [],
+  retryStatusCodes = [],
+  retryFailedRequests = true,
+  onRequestFailure,
+  onRequestSuccess,
+  customFetch = fetch,
+  useStm,
+  eventStore = newInMemoryEventStore({}),
+  credentials,
+}: EmitterConfiguration): Emitter {
+  let idServiceCalled = false;
+  let flushInProgress = false;
+  const usePost = eventMethod.toLowerCase() === 'post';
+  dontRetryStatusCodes = dontRetryStatusCodes.concat([400, 401, 403, 410, 422]);
+
+  function shouldRetryForStatusCode(statusCode: number): boolean {
+    // success, don't retry
+    if (statusCode >= 200 && statusCode < 300) {
+      return false;
+    }
+
+    if (!retryFailedRequests) {
+      return false;
+    }
+
+    // retry if status code among custom user-supplied retry codes
+    if (retryStatusCodes.includes(statusCode)) {
+      return true;
+    }
+
+    // retry if status code *not* among the don't retry codes
+    return !dontRetryStatusCodes.includes(statusCode);
+  }
+
+  async function executeRequest(request: EmitterRequest): Promise<RequestResult> {
+    const fetchRequest = request.toRequest();
+    if (fetchRequest === undefined) {
+      throw new Error('Request is undefined');
+    }
+
+    let payloads = request.getEvents().map((event) => event.getPayload());
+    try {
+      const response = await customFetch(fetchRequest);
+
+      request.cancelTimeoutTimer();
+
+      if (response.ok) {
+        onRequestSuccess?.(payloads);
```

Comment:
Don't completely understand how this would be useful in GTM SS, but have added the `Response` as the second argument in the callbacks!

Also made the callbacks execute in a `setTimeout` to make them non-blocking.

Reply:
With the response we can theoretically work with [sendPixelFromBrowser](https://developers.google.com/tag-platform/tag-manager/server-side/api#sendpixelfrombrowser) which makes more retargeting-style Tags in GTM:SS work with Snowplow as the native event format; you currently need to use GA as the event transport for those Tags to work correctly.